### PR TITLE
msd api to return kubernetes network policy object

### PR DESCRIPTION
**Pull request by abvaidya**
# Description

# Contribution Checklist:
- [x] **The pull request does not introduce any breaking changes**
- [x] **I have read the [contribution guidelines](https://github.com/AthenZ/athenz/blob/master/CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request.**

### CODE REVIEW

File: core/msd/src/main/rdl/KubernetesNetworkPolicy.rdli [hunk 1]
```diff
@@ -0,0 +1,23 @@
+// Copyright The Athenz Authors
+// Licensed under the terms of the Apache version 2.0 license. See LICENSE file for terms.
+
+include "Names.tdl";
+include "KubernetesNetworkPolicy.tdl";
+
+// API endpoint to get the Kubernetes network policy converted from the corresponding MSD policy
+resource KubernetesNetworkPolicyResponse GET "/domain/{domainName}/service/{serviceName}/kubernetesnetworkpolicy" {
```
Comment (rajeshal) [inline core/msd/src/main/rdl/KubernetesNetworkPolicy.rdli:8]: Is it feasible to call one service per request, for an environment with hundreds/thousands of services ?

Comment (rajeshal) [reply core/msd/src/main/rdl/KubernetesNetworkPolicy.rdli:8]: It would be good to have an option to make batch call, another endpoint is also fine.

File: core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+type KubernetesLabelSelectorRequirement Struct {
+    String key; //Label key that the selector applies to
+    String operator; //Operator that is applied to the key. Valid operators are In, NotIn, Exists and DoesNotExist.
+    Array<String> values (optional); //Array of string values. If the operator is In or NotIn, the values array must be non-empty. If the operator is Exists or DoesNotExist, the values array must be empty.
+}
+
+// A label selector is a label query over a set of resources. The result of matchLabels and matchExpressions are ANDed. An empty label selector matches all objects. A null label selector matches no objects.
+type KubernetesLabelSelector Struct {
+    Array<KubernetesLabelSelectorRequirement> matchExpressions; //Array of label selector requirements. The requirements are ANDed.
+    Map<String, String> matchLabels; //Map of label key/value pairs
```
Comment (rajeshal) [inline core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl:17]: `matchExpressions` can accommodate this requirement as well, with a new `isEqual` operator or existing `In`

File: core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+// Request object containing Kubernetes network policy inputs
+type KubernetesNetworkPolicyRequest Struct {
+    String athenzDomainLabel (optional); //Label key name used on pods to identify Athenz domain
+    String athenzServiceLabel; //Label key name used on pods to identify Athenz service
+    String networkPolicyType (optional); //Network policy type, default is vanilla Kubernetes
+    String requestedApiVersion (optional); //Requested network policy apiVersion
+    String networkPolicyNamespace (optional); //Kubernetes namespace for the network policy target
+    KubernetesLabelSelector podSelector (optional); //Kubernetes pod selector for the network policy target
+}
```
Comment (rajeshal) [inline core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl:66]: I think one option is missing - whether to use a combination of namespace selector and label selector to form the k8s policies OR use only the label selector.

File: core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+// Kubernetes network policy egress rule
+type KubernetesNetworkPolicyEgressRule Struct {
+    Array<KubernetesNetworkPolicyPeer> to (optional); //Network policy target, when empty all sources are allowed
+    Array<KubernetesNetworkPolicyPort> ports (optional); //Egress port(s), when empty all ports are allowed
+}
+
+// Kubernetes network policy spec
+type KubernetesNetworkPolicySpec Struct {
+    KubernetesLabelSelector podSelector; //Kubernetes pod selector for the network policy target
```
Comment (rajeshal) [inline core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl:52]: `targetPodSelector` can make the code more readable

Comment (rajeshal) [reply core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl:52]: Please ignore, i understand that you want to make the result in standard k8s network policy json format

File: core/msd/src/main/rdl/KubernetesNetworkPolicy.rdli (hunk 1 above)
Comment (abvaidya) [reply core/msd/src/main/rdl/KubernetesNetworkPolicy.rdli:8]: yeah that could be another endpoint.

File: core/msd/src/main/rdl/KubernetesNetworkPolicy.rdli [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    DomainName domainName; //Name of the domain
+    EntityName serviceName; //Name of the service
+    KubernetesNetworkPolicyRequest request; //Struct representing input options based on the cluster context
```
[the remark this answers is not in the document]
Comment (abvaidya) [reply core/msd/src/main/rdl/KubernetesNetworkPolicy.rdli:11]: Good catch! It wont, I will update it to be POST.

File: core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl (hunk 3 above)
Comment (abvaidya) [reply core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl:66]: Both the fields are optional for that reason. Caller can set either or both or none.

File: core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+// A label selector is a label query over a set of resources. The result of matchLabels and matchExpressions are ANDed. An empty label selector matches all objects. A null label selector matches no objects.
+type KubernetesLabelSelector Struct {
+    Array<KubernetesLabelSelectorRequirement> matchExpressions; //Array of label selector requirements. The requirements are ANDed.
+    Map<String, String> matchLabels; //Map of label key/value pairs
+}
+
+// Kubernetes network policy port range
+type KubernetesNetworkPolicyPort PolicyPort {
```
Comment (rajeshal) [inline core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl:21]: Can't we reuse `TransportPolicyPort` here, like `TransportPolicyProtocol` ?

Comment (abvaidya) [reply core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl:21]: I would prefer to keep the objects separate for api calls even though they have similar members for now. RDL doesnt like to share Enums unlike Struct so couldn't create KubernetesNetworkPolicyProtocol

File: core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl (hunk 3 above)
Comment (rajeshal) [reply core/msd/src/main/rdl/KubernetesNetworkPolicy.tdl:66]: upon further review, noticed this definition uses namespace selector as well, so let me rephrase the previous question - how can user choose between `combination of namespace selector and label selector` and `only the label selector, using athenzServiceLabel` ?

**Review by rajeshal — APPROVED** [review]
👍 LGTM

**Review by havetisyan — APPROVED** [review]
